**TAMSIN ROGERS**

Spring 2021

CS 251: Data Analysis and Visualization

Project 6: Supervised learning

In [139]:
import os
import random
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

plt.style.use(['seaborn-colorblind', 'seaborn-darkgrid'])
plt.rcParams.update({'font.size': 20})

np.set_printoptions(suppress=True, precision=5)

# Automatically reload external modules
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


## Task 3: Preprocess full spam email dataset 

Before you build a Naive Bayes spam email classifier, run the full spam email dataset through your preprocessing code.

Download and extract the full **Enron** emails (*zip file should be ~29MB large*). You should see a base `enron` folder, with `spam` and `ham` subfolders when you extract the zip file (these are the 2 classes).

Run the test code below to check everything over.

### 3a) Preprocess dataset

In [71]:
import email_preprocessor as epp

#### Test `count_words` and `find_top_words`

In [72]:
word_freq, num_emails = epp.count_words()

In [124]:
print(f'You found {num_emails} emails in the datset. You should have found 32625.')

You found 32625 emails in the datset. You should have found 32625.


In [74]:
top_words, top_counts = epp.find_top_words(word_freq)
print(f"Your top 5 words are\n{top_words[:5]}\nand they should be\n['the', 'to', 'and', 'of', 'a']")
print(f"The associated counts are\n{top_counts[:5]}\nand they should be\n[277459, 203659, 148873, 139578, 111796]")

Your top 5 words are
['the', 'to', 'and', 'of', 'a']
and they should be
['the', 'to', 'and', 'of', 'a']
The associated counts are
[277459, 203659, 148873, 139578, 111796]
and they should be
[277459, 203659, 148873, 139578, 111796]


### 3b) Make train and test splits of the dataset

Here we divide the email features into a 80/20 train/test split (80% of data used to train the supervised learning model, 20% we withhold and use for testing / prediction).

In [129]:
features, y = epp.make_feature_vectors(top_words, num_emails)
np.random.seed(0)
x_train, y_train, inds_train, x_test, y_test, inds_test = epp.make_train_test_sets(features, y)

In [130]:
print('Shapes for train/test splits:')
print(f'Train {x_train.shape}, classes {y_train.shape}')
print(f'Test {x_test.shape}, classes {y_test.shape}')
print('\nThey should be:\nTrain (26100, 200), classes (26100,)\nTest (6525, 200), classes (6525,)')

Shapes for train/test splits:
Train (26100, 200), classes (26100,)
Test (6525, 200), classes (6525,)

They should be:
Train (26100, 200), classes (26100,)
Test (6525, 200), classes (6525,)


### 3c) Save data in binary format

It add a lot of overhead to have to run through your raw email -> train/test feature split every time you wanted to work on your project! In this step, you will export the data in memory to disk in a binary format. That way, you can quickly load all the data back into memory (directly in ndarray format) whenever you want to work with it again. No need to parse from text files!

- Use numpy's `save` function to make six files in `.npy` format (e.g. `email_train_x.npy`, `email_train_y.npy`, `email_train_inds.npy`, `email_test_x.npy`, `email_test_y.npy`, `email_test_inds.npy`).

In [131]:
np.save("email_train_x.npy", x_train)
np.save("email_train_y.npy", y_train)
np.save("email_train_inds.npy", inds_train)
np.save("email_test_x.npy", x_test)
np.save("email_test_y.npy", y_test)
np.save("email_test_inds.npy", inds_test)

## Task 4: Naive Bayes Classifier

After finishing your email preprocessing pipeline, implement the one other supervised learning algorithm we we will use to classify email, **Naive Bayes**.

### 4a) Implement Naive Bayes

In `naive_bayes.py`, implement the following methods:
- Constructor
- `train(data, y)`: Train the Naive Bayes classifier so that it records the "statistics" of the training set: class priors (i.e. how likely an email is in the training set to be spam or ham?) and the class likelihoods (the probability of a word appearing in each class — spam or ham).
- `predict(data)`: Combine the class likelihoods and priors to compute the posterior distribution. The predicted class for a test sample is the class that yields the highest posterior probability.
- `accuracy(y, y_pred)`: The usual definition :)


#### Bayes rule ingredients: Priors and likelihood (`train`)

To compute class predictions (probability that a test example belong to either spam or ham classes), we need to evaluate **Bayes Rule**. This means computing the priors and likelihoods based on the training data.

**Prior:** $$P_c = \frac{N_c}{N}$$ where $P_c$ is the prior for class $c$ (spam or ham), $N_c$ is the number of training samples that belong to class $c$ and $N$ is the total number of training samples.

**Likelihood:** $$L_{c,w} = \frac{N_{c,w} + 1}{N_{c} + M}$$ where
- $L_{c,w}$ is the likelihood that word $w$ belongs to class $c$ (*i.e. what we are solving for*)
- $N_{c,w}$ is the total count of **word $w$** in emails that are only in class $c$ (*either spam or ham*)
- $N_{c}$ is the total number of **all words** that appear in emails of the class $c$ (*total number of words in all spam emails or total number of words in all ham emails*)
- $M$ is the number of features (*number of top words*).

#### Bayes rule ingredients: Posterior (`predict`)

To make predictions, we now combine the prior and likelihood to get the posterior:

**Log Posterior:** $$Log(\text{Post}_{i, c}) = Log(P_c) + \sum_{j \in J_i}x_{i,j}Log(L_{c,j})$$

 where
- $\text{Post}_{i,c}$ is the posterior for class $c$ for test sample $i$(*i.e. evidence that email $i$ is spam or ham*). We solve for its logarithm.
- $Log(P_c)$ is the logarithm of the prior for class $c$.
- $x_{i,j}$ is the number of times the jth word appears in the ith email.
- $Log(L_{c,j})$: is the log-likelihood of the jth word in class $c$.

In [132]:
from naive_bayes_multinomial import NaiveBayes

#### Test `train`

In [133]:
num_test_classes = 4
np.random.seed(0)
data_test = np.random.random(size=(100, 6))
y_test = np.random.randint(low=0, high=num_test_classes, size=(100,))

nbc = NaiveBayes(num_classes=num_test_classes)
nbc.train(data_test, y_test)

print(f'Your class priors are: {nbc.class_priors}\nand should be          [0.24 0.26 0.25 0.25].')
print(f'Your class likelihoods shape is {nbc.class_likelihoods.shape} and should be (4, 6).')
print(f'Your likelihoods are:\n{nbc.class_likelihoods}')

test_likelihoods = np.array([[0.15116, 0.18497, 0.17571, 0.1463 , 0.16813, 0.17374],
       [0.16695, 0.17437, 0.15742, 0.16887, 0.15677, 0.17562],
       [0.14116, 0.1562 , 0.19651, 0.17046, 0.17951, 0.15617],
       [0.18677, 0.18231, 0.15884, 0.12265, 0.16755, 0.18187]])
print(f'and should be\n{test_likelihoods}')


Your class priors are: [0.24 0.26 0.25 0.25]
and should be          [0.24 0.26 0.25 0.25].
Your class likelihoods shape is (4, 6) and should be (4, 6).
Your likelihoods are:
[[0.15116 0.18497 0.17571 0.1463  0.16813 0.17374]
 [0.16695 0.17437 0.15742 0.16887 0.15677 0.17562]
 [0.14116 0.1562  0.19651 0.17046 0.17951 0.15617]
 [0.18677 0.18231 0.15884 0.12265 0.16755 0.18187]]
and should be
[[0.15116 0.18497 0.17571 0.1463  0.16813 0.17374]
 [0.16695 0.17437 0.15742 0.16887 0.15677 0.17562]
 [0.14116 0.1562  0.19651 0.17046 0.17951 0.15617]
 [0.18677 0.18231 0.15884 0.12265 0.16755 0.18187]]


#### Test `predict`

In [134]:
num_test_classes = 4
np.random.seed(0)
data_train = np.random.randint(low=0, high=num_test_classes, size=(100, 10))
data_test = np.random.randint(low=0, high=num_test_classes, size=(15, 10))
y_test = np.random.randint(low=0, high=num_test_classes, size=(100,))

nbc = NaiveBayes(num_classes=num_test_classes)
nbc.train(data_train, y_test)
test_y_pred = nbc.predict(data_test)

print(f'Your predicted classes are\n{test_y_pred}\nand should be\n[3 0 3 1 0 1 1 3 0 3 0 2 0 2 1]')

Your predicted classes are
[3 0 3 1 0 1 1 3 0 3 0 2 0 2 1]
and should be
[3 0 3 1 0 1 1 3 0 3 0 2 0 2 1]


### 4b) Spam filtering

Let's start classifying spam email using the Naive Bayes classifier.

- Use `np.load` to load in the train/test split that you created last week.
- Use your Naive Bayes classifier on the Enron email dataset!

**Question 7:** Print out the accuracy that you get on the test set with Naive Bayes. It should be roughly 89%.

In [135]:
import email_preprocessor as ep

In [136]:
train_x = np.load('email_train_x.npy')
train_y = np.load('email_train_y.npy')
test_x = np.load('email_test_x.npy')
test_y = np.load('email_test_y.npy')

nbc = NaiveBayes(num_classes = 2)
nbc.train(x_train, y_train)

y_pred = nbc.predict(test_x)
acc = nbc.accuracy(test_y, y_pred)
print(acc)

0.8924137931034483


### 4c) Confusion matrix

To get a better sense of the errors that the Naive Bayes classifer makes, you will create a confusion matrix. 

- Implement `confusion_matrix` in `naive_bayes.py`.
- Print out a confusion matrix of the spam classification results.

**Debugging guidelines**:
1. The sum of all numbers in your 2x2 confusion matrix should equal the number of test samples (6525).
2. The sum of your spam row should equal the number of spam samples in the test set (3193)
3. The sum of your ham row should equal the number of spam samples in the test set (3332)

In [146]:
confuse = nbc.confusion_matrix(test_y, y_pred)
print(confuse[0])

[[3043  150]
 [ 552 2780]]


**Question 8:** Interpret the confusion matrix, using the convention that positive detection means spam (*e.g. a false positive means classifying a ham email as spam*). What types of errors are made more frequently by the classifier? What does this mean (*i.e. X (spam/ham) is more likely to be classified than Y (spam/ham) than the other way around*)?

**Reminder:** Look back and make sure you are clear on which class indices correspond to spam/ham.

**Answer 8:**
The confusion matrix has the form:

trueneg, falsepos

falseneg, truepos

The classifier classifies spam as ham (false negative = 552) more than ham as spam (false positive = 150).  

### 4d) Investigate the misclassification errors

Numbers are nice, but they may not the best for developing your intuition. Sometimes, you want to see what an misclassification *actually looks like* to help you improve your algorithm. Here, you will take a false positive and a false negative misclassification and retrieve the actual text of the email so see which emails produced the error.

- Determine the index of the **FIRST** false positive and false negative misclassification — i.e. 2 indices in total. Remember to use your `test_inds` array to look up the index of the emails BEFORE shuffling happened.
- Implement the function `retrieve_emails` in `email_preprocessor.py` to return the string of the raw email at the error indices.
- Call your function to print out the two emails that produced misclassifications.

**Question 9:** Does it seem reasonable that each email message was misclassified? Why?

**Answer 9:** 
Yes, the false positive email is ham, but was misclassified as spam.  The false negative email is spam, but was misclassified as ham, which we know because it contains spam words.

In [163]:
falsepos = confuse[1]
falseneg = confuse[2]
truepos = confuse[3]
trueneg = confuse[4]

email = ep.retrieve_emails([inds_test[falsepos[0]]], 'data/enron')
falseposemail = ep.retrieve_emails([inds_test[falsepos[0]]], 'data/enron')
falsenegemail = ep.retrieve_emails([inds_test[falseneg[0]]], 'data/enron')

print("FALSE POS EMAIL (ham classified as spam):")
print(falseposemail)
print("\n")
print("FALSE NEG EMAIL (spam classified as ham):")
print(falsenegemail)

FALSE POS EMAIL (ham classified as spam):
[['subject', 'view', 'and', 'print', 'doorstep', 'reports', 'view', 'reports', 'several', 'project', 'reports', 'or', 'views', 'are', 'available', 'to', 'include', 'all', 'complete', 'and', 'open', 'select', 'the', 'view', 'of', 'choice', 'by', 'clicking', 'view', 'and', 'then', 'go', 'to', 'printing', 'a', 'single', 'project', 'view', 'step', 'expand', 'the', 'project', 'you', 'wish', 'to', 'print', 'by', 'clicking', 'the', 'view', 'expand', 'button', 'select', 'the', 'project', 'name', 'and', 'the', 'applicable', 'observations', 'by', 'left', 'clicking', 'and', 'placing', 'a', 'check', 'by', 'each', 'step', 'left', 'click', 'on', 'file', 'then', 'print', 'select', 'printer', 'then', 'setup', 'and', 'change', 'paper', 'size', 'to', 'legal', 'and', 'orientation', 'to', 'landscape', 'left', 'click', 'ok', 'select', 'view', 'options', 'print', 'view', 'and', 'left', 'click', 'ok', 'make', 'certain', 'to', 'set', 'printer', 'default', 'print', 'op

## Task 5: Comparison with KNN


- Run a similar analysis to what you did with Naive Bayes above. When computing accuracy on the test set, you may want to reduce the size of the test set (e.g. to the first 500 emails in the test set).
- Copy-paste your `confusion_matrix` method into `knn.py` so that you can run the same analysis on a KNN classifier.

In [166]:
from knn import KNN

In [168]:
knn = KNN(num_classes = 2)
knn.train(x_train, y_train)

y_pred = knn.predict(test_x, k=1)
acc = knn.accuracy(test_y, y_pred)
print(acc)

0.5489655172413793


In [171]:
confuse2 = knn.confusion_matrix(test_y, y_pred)
print(confuse2[0])

[[1818 1375]
 [1568 1764]]


In [172]:
falsepos2 = confuse2[1]
falseneg2 = confuse2[2]
truepos2 = confuse2[3]
trueneg2 = confuse2[4]

email2 = ep.retrieve_emails([inds_test[falsepos2[0]]], 'data/enron')
falseposemail2 = ep.retrieve_emails([inds_test[falsepos2[0]]], 'data/enron')
falsenegemail2 = ep.retrieve_emails([inds_test[falseneg2[0]]], 'data/enron')

print("FALSE POS EMAIL (ham classified as spam):")
print(falseposemail2)
print("\n")
print("FALSE NEG EMAIL (spam classified as ham):")
print(falsenegemail2)

FALSE POS EMAIL (ham classified as spam):
[['subject', 'east', 'power', 'project', 'directory', 'a', 'new', 'directory', 'm', 'eprojects', 'directory', 'nahoutrd', 'houston', 'pwr', 'common', 'eprojects', 'has', 'been', 'created', 'to', 'house', 'all', 'of', 'the', 'related', 'documents', 'that', 'will', 'be', 'used', 'to', 'manage', 'the', 'east', 'trading', 'enhancement', 'projects', 'the', 'intent', 'of', 'this', 'directory', 'is', 'to', 'store', 'all', 'of', 'the', 'business', 'requirements', 'documentation', 'and', 'project', 'management', 'for', 'this', 'project', 'as', 'well', 'as', 'all', 'future', 'work', 'on', 'our', 'systems', 'there', 'has', 'been', 'directories', 'set', 'up', 'accenture', 'north', 'east', 'mid', 'west', 'south', 'east', 'and', 'ercot', 'below', 'these', 'main', 'directories', 'there', 'is', 'three', 'directories', 'for', 'risk', 'systems', 'casey', 'stacey', 'settlement', 'systems', 'thresa', 'and', 'trading', 'systems', 'correy', 'a', 'draft', 'scope', 'a

**Question 10:** What accuracy did you get on the test set (potentially reduced in size)?

**Question 11:** How does the confusion matrix compare to that obtained by Naive Bayes (*If you reduced the test set size, keep that in mind*)?

**Question 12:** Briefly describe at least one pro/con of KNN compared to Naive Bayes on this dataset.

**Question 13:** When potentially reducing the size of the test set here, why is it important that we shuffled our train and test set?

**Answer 10:**
I got an accuracy of 0.549

**Answer 11:** 
The KNN confusion matrix is:

trueneg, falsepos

falseneg, truepos

1818 1375

1568 1764

So the classifier misclassifies spam as ham (falseneg = 1568) more than it misclassifies ham as spam (falsepos= 1375). 

**Answer 12:** 

KNN is less accurate and misclassifies more.

**Answer 13:** 

It is important that we shuffled our train and test set in order to distribute the data appropriately.  If the data is in any kind of order and the size is reduced, it will give biased results.

I worked on this project with Rayna Hata & Nicole Matamoros

## Extensions

### 0. Classify your own datasets

- Find datasets that you find interesting and run classification on them using your KNN algorithm (and if applicable, Naive Bayes). Analysis the performance of your classifer.

### 1. Better text preprocessing

- If you look at the top words extracted from the email dataset, many of them are common "stop words" (e.g. a, the, to, etc.) that do not carry much meaning when it comes to differentiating between spam vs. non-spam email. Improve your preprocessing pipeline by building your top words without stop words. Analyze performance differences.

### 2. Feature size

- Explore how the number of selected features for the email dataset influences accuracy and runtime performance.

### 3. Distance metrics
- Compare KNN performance with the $L^2$ and $L^1$ distance metrics

### 4. K-Fold Cross-Validation

- Research this technique and apply it to data and your KNN and/or Naive Bayes classifiers.

### 5. Email error analysis

- Dive deeper into the properties of the emails that were misclassified (FP and/or FN) by Naive Bayes or KNN. What is their word composition? How many words were skipped because they were not in the training set? What could plausibly account for the misclassifications?